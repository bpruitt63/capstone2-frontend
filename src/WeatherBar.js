import React, {useState} from 'react';
import {Collapse} from 'reactstrap';
import './static/styles/WeatherBar.css';


/** Component displayed at top of all pages, showing today's weather 
 * for current location.
 */
function WeatherBar({location, weather, isMobile, units}) {

    const [isOpen, setIsOpen] = useState(false);

    /** Toggles collapsible weather on small screens */
    const toggle = () => setIsOpen(!isOpen);

    /** Gives loading or error message if weather prop is
     * undefined or error string
     */
    if (!weather) return <p className='WeatherBar'>Getting weather</p>;
    if (weather === 'API error') return <p className='WeatherBar'>API Error</p>;
    if (weather === 'Bad location') {
        return (
            <div>
                <p className='WeatherBar'>That place don't exist!</p>
            </div>
        );
    };

    const w = weather.data.data[0];
    const countryDisplay = Number.isInteger(+weather.data.state_code) ? 
                            location.country : weather.data.state_code || weather.data.country_code;
    const ICON_URL = 'https://www.weatherbit.io/static/img/icons/'

    return (
        <>
        {!isMobile &&
            <div className='WeatherBar'>
                <div className='container'>
                    <div className='row'>
                        <div className='col-3 align-self-start'>
                            <img src={`${ICON_URL}${w.weather.icon}.png`} 
                                    alt={w.weather.description}
                                    className='WeatherBarImage' />
                            <p>{w.weather.description}</p>
                        </div>
                        <div className='col-9'>
                            <div className='row'>
                                <div className='row'>
                                    <h5 className='col-8'>Today's weather for {weather.data.city_name} 
                                        , {countryDisplay}.</h5>
                                </div>
                                <div className='col-3 align-self-start'>
                                    <p>High: {w.high_temp}</p>
                                    <p className='test'>Feels like: {w.app_max_temp}</p>
                                </div>
                                <div className='col-3 align-self-start'>
                                    <p>Low: {w.low_temp}</p>
                                    <p>Feels like: {w.app_min_temp}</p>
                                </div>
                                <div className='col-3 align-self-start'>
                                    <p>Precipitation:</p>
                                    <p>{w.pop}%</p>
                                </div>
                                <div className='col-3 align-self-start'>
                                    <p>Wind: {w.wind_spd}
                                        {units === 'i' ? 'mph ' : 'km/h '}
                                        {w.wind_cdir}
                                    </p>
                                    <p>Gust: {w.wind_gust_spd}
                                        {units === 'i' ? 'mph ' : 'km/h '}
                                    </p>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        }
        {isMobile &&
            <div className='WeatherBar'>
                <div className='container'>
                    <div className='row'>
                        <div className='col-4'>
                            <img src={`${ICON_URL}${w.weather.icon}.png`} 
                                    alt={w.weather.description}
                                    className='WeatherBarImage' />
                            {isOpen && 
                                <p>{w.weather.description}</p>
                            }
                        </div>
                        <div className='col-4 align-self-center'>
                            <p><b>{weather.data.city_name} 
                                    , {weather.data.state_code || location.country}</b>
                            </p>
                            <button className='dropdown-toggle WeatherBarToggle' onClick={toggle}>
                                {isOpen ? 'See Less' : 'See More'}
                            </button>
                        </div>
                        <div className='col-4 align-self-end'>
                            <p>High
                                <span><br/>{w.high_temp}</span>
                            </p>
                            {isOpen &&
                                <p>Feels like
                                    <span><br/>{w.app_max_temp}</span>
                                </p>
                            }
                        </div>
                    </div>
                    <Collapse isOpen={isOpen}>
                        <div className='row second'>
                            <p className='col-4 align-self-start'>Precipitation
                                <span><br/>{w.pop}%</span>
                            </p>
                            <div className='col-4 align-self-start'>
                                <p>Wind
                                    <span><br/>{w.wind_spd}
                                        {units === 'i' ? 'mph ' : 'km/h '}
                                        {w.wind_cdir}
                                    </span>
                                </p>
                                <p>Gust
                                    <span><br/>{w.wind_gust_spd}
                                        {units === 'i' ? 'mph ' : 'km/h '}
                                    </span>
                                </p>
                            </div>
                            <div className='col-4 align-self-start'>
                                <p>Low
                                    <span><br/>{w.low_temp}</span>
                                </p>
                                <p>Feels like
                                    <span><br/>{w.app_min_temp}</span>
                                </p>
                            </div>
                        </div>
                    </Collapse>
                </div>
            </div>
        }
        </>
    );
};

export default WeatherBar;