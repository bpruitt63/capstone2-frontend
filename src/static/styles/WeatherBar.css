.WeatherBar {
    background-color: rgb(162, 224, 240);
    padding-top: 15px;
    margin-bottom: 15px;
}

.WeatherBar div div {
    background-color: rgb(162, 224, 240);
    border: none;
}

.WeatherBar div div div img {
    width: 100px;
    margin: -10px 0px;
}

h5 {
    padding-bottom: 10px;
    text-align: left;
}

.card-body {
    text-align: left;
    margin: -20px 0px -10px 0px;
}

.WeatherBarToggle {
    background-color: rgb(162, 224, 240);
    border: none;
}



