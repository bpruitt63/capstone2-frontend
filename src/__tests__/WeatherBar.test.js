import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import WeatherBar from '../WeatherBar';

const weather = {data: {city_name: 'testCity',
                        state_code: 'OH',
                        data: [{wind_spd: 10, wind_gust_spd: 15, wind_cdir: 'N',
                        high_temp: 80, app_max_temp: 90, low_temp: 60, 
                        app_min_temp: 55, pop: 15,
                        weather: {
                            icon: 'none',
                            description: 'test description'
                        } }]}};

test('renders', () => {
    render(<WeatherBar weather={weather}/>);
});

test('matches snapshot', () => {
    const {asFragment} = render(<WeatherBar weather={weather}/>);
    expect(asFragment()).toMatchSnapshot();
});

test('renders non mobile', () => {
    render(<WeatherBar weather={weather} isMobile={false}/>);
    expect(screen.getByText(/Today's weather for testCity, OH./)).toBeInTheDocument;
    expect(screen.queryByText(/See More/)).not.toBeInTheDocument;
});

test('renders mobile', () => {
    render(<WeatherBar weather={weather} isMobile={true}/>);
    expect(screen.queryByText(/Today's weather for testCity, OH./)).not.toBeInTheDocument;
    expect(screen.getByText(/testCity, OH/)).toBeInTheDocument;
    expect(screen.getByText(/See More/)).toBeInTheDocument;
});

test('toggles mobile collapse', () => {
    render(<WeatherBar weather={weather} isMobile={true}/>);
    expect(screen.queryByText(/test description/)).not.toBeInTheDocument;
    expect(screen.getByText(/See More/)).toBeInTheDocument;
    const toggle = screen.getByText(/See More/);
    fireEvent.click(toggle);
    expect(screen.getByText(/test description/)).toBeInTheDocument;
    fireEvent.click(toggle);
    expect(screen.queryByText(/test description/)).not.toBeInTheDocument;
});

test('renders with error', () => {
    render(<WeatherBar weather={'Bad location'}/>);
    expect(screen.getByText(/That place don't exist!/)).toBeInTheDocument;
});