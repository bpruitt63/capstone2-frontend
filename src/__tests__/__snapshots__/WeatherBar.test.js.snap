// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="WeatherBar"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-3 align-self-start"
        >
          <img
            alt="test description"
            class="WeatherBarImage"
            src="https://www.weatherbit.io/static/img/icons/none.png"
          />
          <p>
            test description
          </p>
        </div>
        <div
          class="col-9"
        >
          <div
            class="row"
          >
            <div
              class="row"
            >
              <h5
                class="col-8"
              >
                Today's weather for testCity, OH.
              </h5>
            </div>
            <div
              class="col-3 align-self-start"
            >
              <p>
                High: 80
              </p>
              <p
                class="test"
              >
                Feels like: 90
              </p>
            </div>
            <div
              class="col-3 align-self-start"
            >
              <p>
                Low: 60
              </p>
              <p>
                Feels like: 55
              </p>
            </div>
            <div
              class="col-3 align-self-start"
            >
              <p>
                Precipitation:
              </p>
              <p>
                15%
              </p>
            </div>
            <div
              class="col-3 align-self-start"
            >
              <p>
                Wind: 10km/h N
              </p>
              <p>
                Gust: 15km/h 
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
